### DISCUSSION TITLE
Why Predicted Heat maps trained using efficient_ad models look strange.

### DISCUSSION THREAD

**Discussion by Xuyilongs** [open-edge-platform/anomalib#1647] [Q&A] [answered]
![000](https://github.com/openvinotoolkit/anomalib/assets/106374953/341283fe-7689-4bae-bab2-aad430ed8a1c)
This is the result picture I got when I trained the efficient_ad model using the default data set format, why the predicted heat map here is strange, all green.
![image](https://github.com/openvinotoolkit/anomalib/assets/106374953/2c36b615-b869-4bc9-bfb5-d9ec0cb3c886)
Why is the anomaly_maps data always the same value 0.4975?
![000](https://github.com/openvinotoolkit/anomalib/assets/106374953/4fa36a4a-d324-4c4e-a2c7-328539be00bf)
But on the same data set, using other models such as fastflow, this is what the predicted heat map looks like.

**Comment by blaz-r** [comment]
Hi, this is indeed quite unusual. I don't remember if I've ever visualized the EfficientAD anomaly maps, so we'll need to check that.

Since the binary segmentation map is still valid, I think that the value difference between anomalous and normal region is just so small in this case, making it impossible to see.  This is probably due invalid normalization values or approach. Are all the masks like this?

It is possible that since the map normalization is part of the model, adaptive threshold value that is calculated on maps causes minmax to scale the image in this way (since this threshold is also later used in [minmax normalization](https://github.com/openvinotoolkit/anomalib/blob/main/src/anomalib/utils/callbacks/min_max_normalization.py#L97)).

**Comment by abc-125** [comment]
Hello, I am not sure if this is the case, but please check that you have `normalization: none` in your config file. EfficientAD handles normalization inside the model, so using the default config from a different model might be the problem.

**Reply by blaz-r** [reply]
Yeah, that might be a bit problematic then 😅 as the default normalization method for EfficientAD is set to minmax in config:
https://github.com/openvinotoolkit/anomalib/blob/2e09314a2d5ab19321806d0789fc97d369254582/src/anomalib/models/efficient_ad/config.yaml#L21-L30

**Reply by abc-125** [reply, marked as answer]
I meant normalization for the `dataset`, it can be set to `imagenet`, which would be a problem in the case of EfficientAD 😅
https://github.com/openvinotoolkit/anomalib/blob/2e09314a2d5ab19321806d0789fc97d369254582/src/anomalib/models/efficient_ad/config.yaml#L12

**Reply by blaz-r** [reply]
Oh, that is correct. But since the image above seems to have correct binary mask I think that the model actually learns fine, but the visualization is not right. Can you confirm this @Xuyilongs or is the model not learning at all?

**Reply by abc-125** [reply]
@Xuyilongs can you also please share your `config.yaml` file?

**Reply by Xuyilongs** [reply]
Thanks for your answer, here is my config.yam

dataset:
  name: mvtec
  format: mvtec
  path: ./datasets/MVTec
  category: bottle
  task: segmentation  # 可选：classification，detection，segmentation
  train_batch_size: 1
  eval_batch_size: 16
  num_workers: 8
  image_size: 256 # dimensions to which images are resized (mandatory)
  center_crop: null # dimensions to which images are center-cropped after resizing (optional)
  normalization: none # data distribution to which the images will be normalized: [none, imagenet]
  transform_config:
    train: null
    eval: null
  test_split_mode: from_dir # options: [from_dir, synthetic]
  test_split_ratio: 0.2 # fraction of train images held out testing (usage depends on test_split_mode)
  val_split_mode: same_as_test # options: [same_as_test, from_test, synthetic]
  val_split_ratio: 0.5 # fraction of train/test images held out for validation (usage depends on val_split_mode)

model:
  name: efficient_ad
  teacher_out_channels: 384
  model_size: small # options: [small, medium]
  lr: 0.0001
  weight_decay: 0.00001
  padding: false
  pad_maps: true # relevant for "padding: false", see EfficientAd in lightning_model.py
  # generic params
  normalization_method: min_max # options: [null, min_max, cdf]

metrics:
  image:
    - F1Score
    - AUROC
  pixel:
    - F1Score
    - AUROC
  threshold:
    method: adaptive #options: [adaptive, manual]
    manual_image: null
    manual_pixel: null

visualization:
  show_images: False # show images on the screen
  save_images: True # save images to the file system
  log_images: False # log images to the available loggers (if any)
  image_save_path: null # path to which images will be saved
  mode: full # options: ["full", "simple"]

project:
  seed: 42
  path: ./results

logging:
  logger: [] # options: [comet, tensorboard, wandb, csv] or combinations.
  log_graph: false # Logs the model graph to respective logger.

optimization:
  export_mode: null # options: torch, onnx, openvino
# PL Trainer Args. Don't add extra parameter here.
trainer:
  enable_checkpointing: true
  default_root_dir: null
  gradient_clip_val: 0
  gradient_clip_algorithm: norm
  num_nodes: 1
  devices: 1
  enable_progress_bar: true
  overfit_batches: 0.0
  track_grad_norm: -1
  check_val_every_n_epoch: 1
  fast_dev_run: false
  accumulate_grad_batches: 1
  max_epochs: 20
  min_epochs: null
  max_steps: 70000
  min_steps: null
  max_time: null
  limit_train_batches: 1.0
  limit_val_batches: 1.0
  limit_test_batches: 1.0
  limit_predict_batches: 1.0
  val_check_interval: 1.0
  log_every_n_steps: 50
  accelerator: auto # <"cpu", "gpu", "tpu", "ipu", "hpu", "auto">
  strategy: null
  sync_batchnorm: false
  precision: 32
  enable_model_summary: true
  num_sanity_val_steps: 0
  profiler: null
  benchmark: false
  deterministic: false
  reload_dataloaders_every_n_epochs: 0
  auto_lr_find: false
  replace_sampler_ddp: true
  detect_anomaly: false
  auto_scale_batch_size: false
  plugins: null
  move_metrics_to_cpu: false
  multiple_trainloader_mode: max_size_cycle

**Comment by Xuyilongs** [comment]
@blaz-r @abc-125 Thank you for your help. You guys have been friendly.

**Reply by SilverStarCoder** [reply]
Hi, my heatmaps are also looking like yours, even after I set normalization to none. Is there anything else you changed?

**Reply by abc-125** [reply]
It looks like this [PR](https://github.com/openvinotoolkit/anomalib/pull/2130) should fix it, you need this part in particular:
```
def on_validation_epoch_start(self, trainer: Trainer, pl_module: AnomalyModule) -> None:
  """Call when the validation epoch begins."""
  del trainer  # `trainer` variable is not used.
  
  if hasattr(pl_module, "normalization_metrics"):
      pl_module.normalization_metrics.reset()
  ```

**Reply by watertianyi** [reply]
Which file is normalization_method in? How to set normalization_method=none in TorchInferencer during inference?
